File: code/SoundPlayground/SoundPlayground/Parser/AbstractSyntaxTree/MusicRepeatNode.cs
using System.Collections.Generic;
using SoundPlayground.Core;

namespace SoundPlayground.Parser.AbstractSyntaxTree
{
    public class MusicRepeatNode : MusicNode
    {
        public MusicNode Expression { get; set; }

        public int Count { get; set; }

        public MusicRepeatNode ( MusicNode expression, int count ) {
            Expression = expression;
            Count = count;
        }

        public override IEnumerable<Note> GetCommands ( Context context ) {
            for ( int i = 0; i < Count; i++ ) {
                var ctx = context.Fork();

                try {
                    foreach ( Note note in Expression.GetCommands( ctx ) ) {
                        yield return note;
                    }
                } finally {
                    context.Join( ctx );
                }
            }
        }
    }
}

File: code/SoundPlaygroundPy/parser/abstract_syntax_tree/music_repeat_node.py
from .music_node import MusicNode

class MusicRepeatNode( MusicNode ):
    def __init__ ( self, expression, count ):
        super().__init__()

        self.expression = expression
        self.count = count
    
    def get_events ( self, context ):
        for i in range( self.count ):
            ctx = context.fork()

            try:
                for event in self.expression.eval( ctx ):
                    yield event
            finally:
                context.join( ctx )
